File: monkey/token/token.go
package token

type TokenType string

// トークン
type Token struct {
	Type    TokenType
	Literal string
}

const (
	ILLEGAL = "ILLIGAL"
	EOF     = "EOF"

	IDENT = "INDENT"
	INT   = "INT"

	ASSIGN   = "="
	PLUS     = "+"
	MINUS    = "-"
	ASTERISK = "*"
	SLASH    = "/"
	BANG     = "!"

	// ==
	EQ = "=="
	// !=
	NOT_EQ = "!="

	COMMA     = ","
	SEMICOLON = ";"

	// (
	LPAREN = "("
	// )
	RPAREN = ")"
	// {
	LBRACE = "{"
	// }
	RBRACE = "}"

	// >
	LT = "<"
	// >
	GT = ">"

	// fn
	FUNCTION = "FUNCTION"
	// let
	LET = "LET"

	// true
	TRUE = "true"
	// false
	FALSE = "false"
	// if
	IF = "if"
	// else
	ELSE = "else"
	// return
	RETURN = "return"
)

func New(tokenType TokenType, ch byte) Token {
	return Token{
		Type:    tokenType,
		Literal: string(ch),
	}
}

//  識別子
var keywords = map[string]TokenType{
	"fn":     FUNCTION,
	"let":    LET,
	"if":     IF,
	"else":   ELSE,
	"return": RETURN,
	"true":   TRUE,
	"false":  FALSE,
}

func LookupIdent(ident string) TokenType {
	if tok, ok := keywords[ident]; ok {
		return tok
	} else {
		return IDENT
	}
}

File: monkey-rs/src/token.rs
#[derive(Debug, Hash, PartialEq, Eq, Clone)]
pub enum TokenType {
    ILLIGAL,
    EOF,
    // 識別子
    IDENT,
    // 整数
    INT,

    // 演算子
    PLUS,
    MINUS,
    ASTERLISK,
    SLASH,
    // !
    BANG,

    // <
    LT,
    // >
    GT,
    // ==
    EQ,
    // !=
    NotEq,

    // デリミタ
    COMMA,
    SEMICOLON,

    LPAREN,
    RPAREN,
    LBRACE,
    RBRACE,

    // fn
    FUNCTION,
    LET,
    RETURN,
    ASSIGN,

    // true
    TRUE,
    // false
    FALSE,
    // if
    IF,
    // else
    ELSE,
}

#[derive(Debug, Clone)]
pub struct Token {
    pub token_type: TokenType,
    pub literal: String,
}

impl Token {
    pub fn new(token_type: TokenType, literal: String) -> Self {
        Self {
            token_type,
            literal,
        }
    }
}

pub fn lookup_ident(ident: String) -> TokenType {
    match ident.as_str() {
        "fn" => TokenType::FUNCTION,
        "let" => TokenType::LET,
        "if" => TokenType::IF,
        "else" => TokenType::ELSE,
        "True" => TokenType::TRUE,
        "false" => TokenType::FALSE,
        "return" => TokenType::RETURN,
        _ => TokenType::IDENT,
    }
}
